runtime/common/src/precompile/schedule_call.rs

```diff
@@ -155,37 +161,97 @@ impl<AccountId, AddressMapping, Scheduler, ChargeTransactionPayment, Call, Origi
 				)
 				.into();
 
-				let delay = DispatchTime::After(min_delay);
-				let origin = Origin::root().caller().clone();
-
 				let current_id = EvmSchedulerNextID::get();
 				let next_id = current_id
 					.checked_add(1)
 					.ok_or_else(|| ExitError::Other("Scheduler next id overflow".into()))?;
 				EvmSchedulerNextID::set(&next_id);
 
-				let task_address = Scheduler::schedule_named(
-					Encode::encode(&(&"ScheduleCall", current_id)),
-					delay,
+				// task_id = id + sender + reserve_fee
+				let task_id = join_task_id(Encode::encode(&(&"ScheduleCall", current_id)), from, _fee.into());
+				Scheduler::schedule_named(
+					task_id.clone(),
+					DispatchTime::After(min_delay),
 					None,
 					0,
-					origin,
+					Origin::root().caller().clone(),
 					call,
 				)
-				.map_err(|_| ExitError::Other("Scheduler failed".into()))?;
+				.map_err(|_| ExitError::Other("Schedule failed".into()))?;
+
+				Ok((ExitSucceed::Returned, task_id, 0))
+			}
+			Action::Cancel => {
+				let from = input.evm_address_at(1)?;
+				let task_id = input.bytes_at(2 * PER_PARAM_BYTES, 96)?;
+
+				debug::debug!(
+					target: "evm",
+					"cancel call: from: {:?}, task_id: {:?}",
+					from,
+					task_id,
+				);
+
+				// task_id = id + sender + reserve_fee
+				let (_, sender, fee) = split_task_id(&task_id);
+				ensure!(sender == from, ExitError::Other("NoPermission".into()));
+
+				#[cfg(not(feature = "with-ethereum-compatibility"))]
+				{
+					// unreserve the transaction fee for gas_limit
+					let from_account = AddressMapping::get_account_id(&from);
+					ChargeTransactionPayment::unreserve_fee(&from_account, fee.into());
+				}
+
+				Scheduler::cancel_named(task_id).map_err(|_| ExitError::Other("Cancel schedule failed".into()))?;
+
+				Ok((ExitSucceed::Returned, vec![], 0))
+			}
+			Action::Reschedule => {
+				let from = input.evm_address_at(1)?;
+				let task_id = input.bytes_at(2 * PER_PARAM_BYTES, 96)?;
+				let min_delay = input.u32_at(3)?;
+
+				debug::debug!(
+					target: "evm",
+					"reschedule call: from: {:?}, task_id: {:?}, min_delay: {:?}",
+					from,
+					task_id,
+					min_delay,
+				);
+
+				// task_id = id + sender + reserve_fee
+				let (_, sender, _) = split_task_id(&task_id);
+				ensure!(sender == from, ExitError::Other("NoPermission".into()));
 
-				Ok((ExitSucceed::Returned, vec_u8_from_tuple(task_address), 0))
+				let delay = DispatchTime::After(min_delay);
+				Scheduler::reschedule_named(task_id, delay).map_err(|e| {
+					let err_msg: &str = e.into();
+					ExitError::Other(err_msg.into())
+				})?;
+
+				Ok((ExitSucceed::Returned, vec![], 0))
 			}
 		}
 	}
 }
 
-fn vec_u8_from_tuple(task_address: TaskAddress<BlockNumber>) -> Vec<u8> {
-	let mut be_bytes_0 = [0u8; 32];
-	U256::from(task_address.0).to_big_endian(&mut be_bytes_0[..]);
+fn join_task_id(id: Vec<u8>, sender: H160, fee: Balance) -> Vec<u8> {
+	let mut be_bytes = [0u8; 96];
```

Comment:
you can do `Encode::encode((id, sender, Compact(fee)))`, and `Decode::<(Vec<u8>, H160, Compact<Balance>)>::decode(input)`

Also use compact to save some bytes